gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/jdbc/authentication/LDAPAuthenticationSchemeImpl.java

```diff
@@ -724,198 +738,343 @@ private String getSearchAuthPwd() throws Exception {
 	}
 // GemStone changes BEGIN
 
-        /**
-         * Search for the given DN for any LDAP group members. The DN can
-         * correspond to a user's UID entry or another LDAP group itself.
-         * LDAP server bind may or not be anonymous.
-         *
-         * @exception NamingException
-         *              on failure to retrieve the given DN
-         */
-        private void resolveDNForGroup(DirContext ctx, String baseDN,
-            String searchFilter, String[] searchAttributes, boolean topLevel,
-            String group, List<String> groupMembers) throws NamingException {
 
-          NamingEnumeration<SearchResult> results;
-          // Construct Search Filter
-          SearchControls ctls = new SearchControls();
-          if (topLevel) {
-            // Setup an LDAP subtree search scope
-            ctls.setSearchScope(SearchControls.SUBTREE_SCOPE);
-          } else {
-            // Setup an LDAP single object search scope
-            ctls.setSearchScope(SearchControls.OBJECT_SCOPE);
+  /**
+   * Search for the LDAP group's DN in the LDAP server. LDAP server bind
+   * may or not be anonymous.
+   * <p>
+   * If the admin does not want us to do anonymous bind/search, then we
+   * must have been given principal/credentials in order to successfully
+   * bind to perform the user's DN search.
+   *
+   * @throws NamingException on failure to retrieve the LDAP group's base DN
+   */
+  public Set<String> getLDAPGroupMembers(String ldapGroup)
+    throws Exception {
+    ArrayList<String> groupMembers = new ArrayList<String>();
+    // Return all the unique members collected for the group
+    this.recurseGroups(ldapGroup, groupMembers, null);
+    return new UnifiedSet<>(groupMembers);
+  }
+
+
+
+
+
+  /**
+   * Search for the all the Groups the user is member of, the LDAP server. LDAP server bind
+   * may or not be anonymous.
+   * <p>
+   * If the admin does not want us to do anonymous bind/search, then we
+   * must have been given principal/credentials in order to successfully
+   * bind to perform the user's DN search.
+   *
+   * @throws NamingException on failure to retrieve the LDAP group's base DN
+   */
+  public Set<String> getLdapGroupsOfUser(String uid)
+    throws Exception {
+
+    DirContext ctx = getDirContext(uid);
+    String attributesNeeded[] = Arrays.copyOf(attrDN, attrDN.length + 1);
+    attributesNeeded[attributesNeeded.length - 1] = memberOfAttr;
+    NamingEnumeration results = getUserInformation(uid, ctx, attributesNeeded);
+    // If we did not find anything then login failed
+    if (results == null || !results.hasMore())
+      throw new NameNotFoundException();
+
+    SearchResult result = (SearchResult) results.next();
+    ArrayList<String> ldapGroups = new ArrayList<String>();
+    Attribute memberOfAttr = result.getAttributes().get("memberOf");
```

Comment:
This memberOf dependency does not look quite right (though in some setups it possible that they are setup both ways). In many cases it would be expected that only group definitions will have the members listed out. I think the proper way is to use a search pattern like: "(&(|(objectClass=group)(objectClass=groupOfNames)(objectClass=groupOfMembers)(objectClass=groupOfUniqueNames))(|(member=%DN%)(member=%UID%)(uniqueMember=%DN%)(uniqueMember=%UID%))" where DN and UID for the user have been obtained by lookup for the user first (like in getDNFromUID). Ideally this pattern can be created from existing searchGroupFilter and searchGroupAttributes by some chopping and merging instead of a new property.

Reply:
In case of customer, they have AD groups which has this memberOf attribute, so it is a short circuit. If this attribute is absent , the code already does the above ( this was already present & have utilized that). what else do you think is missing?